get_lutron_cert.py

```diff
@@ -244,6 +244,10 @@ def write_json(self, obj):
     root_text = message["Body"]["SigningResult"]["RootCertificate"]
     with open(CA_CERT_NAME, "wb") as f:
         f.write(root_text.encode("ASCII"))
+        certs = tls_socket.get_peer_cert_chain()
```

Comment:
Here's caseta.crt:
```❯ openssl x509 -in ./caseta.crt -noout -text                                                                                                                   ─╯
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
    Signature Algorithm: ecdsa-with-SHA256
        Issuer: C=US, ST=Pennsylvania, L=Coopersburg, O=Lutron Electronics Co., Inc., CN=RadioRa3Processor247625112E10
        Validity
            Not Before: Mar  1 00:00:00 2019 GMT
            Not After : Feb  5 00:00:00 2119 GMT
        Subject: CN=get_lutron_cert.py/1.3.6.1.4.1.40073.1.2=000000000000/1.3.6.1.4.1.40073.1.3=get_lutron_cert.py
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:ab:94:7b:db:49:e2:23:95:7f:9f:a2:bf:cb:2e:
                    d8:a8:47:ce:5d:c6:af:f3:9f:94:4e:90:95:bd:29:
                    69:79:20:76:08:92:af:4e:67:af:70:93:fe:ff:48:
                    a6:b7:0b:4c:1c:d4:14:2d:c0:c0:08:87:3c:45:34:
                    11:6d:f4:ea:9f:51:85:cd:1c:70:83:f1:4c:dd:d8:
                    3d:f9:2b:32:d8:6c:38:c2:f6:1b:f9:04:ae:6a:91:
                    5a:7d:c9:5f:85:57:97:0d:80:5e:fa:68:5d:c5:ab:
                    41:12:a0:f4:8c:f5:e5:51:2e:d1:3d:e5:d2:cc:11:
                    7b:a2:a4:29:94:5f:2b:1d:07:29:29:9c:14:27:27:
                    f3:ed:29:e5:23:7f:59:66:31:5b:cf:52:f8:97:0e:
                    62:e4:f7:a6:62:87:31:f2:82:6e:84:32:82:f3:bb:
                    5b:cf:5f:76:8d:6b:91:ef:ea:4a:de:fd:af:8a:fe:
                    d2:bc:9c:92:e0:5a:be:22:d4:57:07:22:1c:83:45:
                    59:ef:58:7d:b0:8f:56:2b:3c:35:7c:37:7d:58:6e:
                    f6:10:c0:13:6c:45:62:96:c5:6d:80:99:4a:82:35:
                    93:f8:b3:e6:eb:af:54:7e:5d:19:50:61:17:12:3c:
                    32:64:96:0e:19:fc:b5:79:ee:55:ae:cd:68:bc:39:
                    35:8f
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
    Signature Algorithm: ecdsa-with-SHA256
         30:44:02:20:0c:65:ed:66:bf:a7:0b:9c:02:42:22:3a:55:8a:
         08:01:2b:88:81:5a:64:18:30:32:0b:22:56:da:55:eb:0c:ab:
         02:20:31:7d:a4:f6:5b:59:b6:f1:dc:09:b2:ae:d6:6d:a4:5e:
         72:01:bc:3d:3e:4c:d5:c6:77:fe:fd:a8:8e:db:55:81
```


Here's caseta-bridge.crt
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
    Signature Algorithm: ecdsa-with-SHA256
        Issuer: C=US, ST=Pennsylvania, L=Coopersburg, O=Lutron Electronics Co., Inc., CN=RadioRa3Processor247625112E10
        Validity
            Not Before: Mar  1 00:00:00 2019 GMT
            Not After : Feb  5 00:00:00 2119 GMT
        Subject: C=US, ST=Pennsylvania, L=Coopersburg, O=Lutron Electronics Co., Inc., CN=RadioRa3Processor247625112E10
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:c8:9f:ff:8f:3a:61:52:99:e1:b9:c4:af:77:6a:
                    0a:77:ed:22:9e:3e:c0:92:5d:9b:20:a0:ef:42:06:
                    4f:8d:e3:ba:25:aa:d4:db:c8:d7:0b:84:11:d6:34:
                    35:d9:96:bc:26:b0:b3:e1:64:64:36:e2:bd:1d:e0:
                    64:ca:84:92:65
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment, Data Encipherment, Key Agreement, Certificate Sign, CRL Sign
            X509v3 Basic Constraints: critical
                CA:TRUE
    Signature Algorithm: ecdsa-with-SHA256
         30:46:02:21:00:ed:29:9d:af:0e:9f:95:37:12:d7:72:a6:3c:
         e9:13:19:bf:b4:16:f3:d1:2f:df:76:aa:13:5d:f0:bb:bc:2a:
         17:02:21:00:eb:9c:5e:f5:26:e6:d1:8d:19:6d:21:d1:04:2b:
         07:44:0b:65:7b:14:df:95:12:1e:f5:e1:1a:bc:0d:2c:00:33
```

All that's fine.  Here's the issue - the server cert you hit on port 8081:
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 32641 (0x7f81)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=US, ST=Ohio, L=Cincinnati, O=Lutron Electronics Co., Inc., CN=radioRa3-processors-027-4725-46
        Validity
            Not Before: Dec 28 17:40:12 2021 GMT
            Not After : Dec 28 17:40:12 2041 GMT
        Subject: C=US, ST=Pennsylvania, O=Lutron Electronics Co., Inc., L=Coopersburg, CN=radiora3-247625112e10-server
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:e4:81:9e:07:93:d3:76:91:d0:32:2c:49:11:03:
                    d7:d9:41:cf:1f:bd:59:3b:28:ba:9d:12:ad:17:3e:
                    9d:60:eb:2d:db:d9:27:7e:2b:53:b8:21:ea:de:bf:
                    d1:4b:65:b3:ff:a7:57:b5:ac:13:91:83:d5:c1:37:
                    a9:4d:05:40:b8
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Basic Constraints:
                CA:FALSE
            X509v3 Subject Key Identifier:
                92:61:7B:8F:4E:09:0B:32:B6:9F:B7:4E:11:D2:F7:D9:F3:D7:23:E4
            X509v3 Authority Key Identifier:
                keyid:BA:4D:E5:E0:11:C3:3F:41:78:70:A2:92:17:1B:FF:DB:71:57:3D:9C
                DirName:/C=US/ST=Pennsylvania/L=Coopersburg/O=Lutron Electronics Co., Inc./CN=radioRa3-devices
                serial:1E:37:33:F2:03:8A:84:F3:73:4D:D9:B0:DD:FD:3C:43:2F:AC:2C:CD

            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Server Authentication
            X509v3 Subject Alternative Name:
                DNS:radiora3-247625112e10-server
    Signature Algorithm: sha256WithRSAEncryption
         0a:a2:21:52:a2:bc:0e:a3:44:63:45:3c:8a:e4:31:d6:3a:9f:
         1d:4b:b9:5a:e0:7f:e6:9e:38:7b:ee:5e:73:a5:83:8e:8d:2a:
         e5:43:cd:80:00:d6:90:c1:0d:cc:6d:a1:8c:ba:96:f9:83:ca:
         43:30:b7:2f:7f:ed:ea:e8:af:5f:8a:2f:61:f7:e7:82:92:b2:
         c2:75:07:f5:e8:32:27:62:a7:ce:b9:19:6a:c0:27:cd:a3:1c:
         7e:43:d4:88:31:e1:a0:3b:6a:6f:90:74:df:e6:3d:b7:9d:59:
         93:06:a4:88:55:d3:5b:32:01:6c:6b:b9:e6:df:44:26:1f:23:
         53:4b:f9:21:d7:b7:69:8b:93:17:d8:8c:a8:5b:8a:dd:bf:cd:
         d9:ec:35:da:5f:39:c7:1c:6d:2b:82:d8:a9:9d:34:07:ec:a0:
         0a:2d:7f:fe:06:d8:1f:66:0e:ab:e4:6a:66:10:f2:17:d6:c5:
         e6:aa:91:73:9a:62:6d:b1:56:79:70:62:e7:03:cb:92:33:89:
         94:d1:2a:5d:59:99:ef:fb:9b:85:3e:ac:47:01:89:51:4e:48:
         dc:e5:03:68:63:b3:f7:21:03:78:76:2b:d8:76:30:2d:f3:2e:
         c0:3f:8d:a4:e9:84:21:78:cd:62:7e:37:c3:fb:82:39:5f:bb:
         9b:ed:4c:3c:a4:11:b9:cb:1a:d6:e8:21:45:e0:89:99:5f:14:
         a0:5e:17:74:b0:72:ea:39:6b:33:fb:37:32:5d:ea:85:dd:65:
         84:cd:74:dd:95:00:3e:4b:80:59:42:32:71:34:b6:a8:c8:0f:
         88:33:09:3f:96:64:d1:52:fb:6a:3e:63:8e:63:0e:e8:47:37:
         81:1f:6d:8e:ac:c5:bb:69:34:32:5e:58:5a:9e:1d:ca:c8:73:
         e7:8a:58:d2:f2:8c:79:23:35:fe:18:88:36:5c:7a:5a:04:30:
         84:dc:7b:2d:94:47:3e:32:c5:f7:2d:41:9b:d3:b6:45:5f:04:
         d2:3b:82:e7:87:14:e5:d6:a9:7a:e5:b9:6a:b0:31:20:79:5e:
         71:6e:77:24:bc:76:8a:45:11:20:ee:cc:2c:cc:11:41:5e:5f:
         4d:2f:f9:c8:10:c0:18:13:bc:dc:9c:cd:7d:e7:24:fd:b6:1f:
         b2:c5:76:e3:22:4c:5c:c2:93:0e:6a:09:f0:c4:3d:c4:1a:5b:
         30:d0:6d:2a:86:c5:9a:a5:8d:83:ae:15:e9:8e:85:b5:90:de:
         14:5e:5c:22:de:26:ac:0c:34:f7:01:32:0d:b5:56:ac:43:b3:
         fc:fb:7f:81:f4:0c:15:7c:f6:a3:7d:04:1f:64:de:c0:80:0d:
         4c:32:d4:ce:a0:e6:af:05
```

So the client doesn't trust the server, because it doesn't know it.

Here's showcerts:
```
❯  openssl s_client -showcerts -connect 192.168.0.42:8081 < /dev/null                                                    ─╯
CONNECTED(00000003)
depth=4 C = US, ST = PA, L = CB, O = Lutron Electronics Inc, OU = Lutron Electronics Inc, CN = lutron-root
verify error:num=19:self signed certificate in certificate chain
verify return:0
---
Certificate chain
 0 s:/C=US/ST=Pennsylvania/O=Lutron Electronics Co., Inc./L=Coopersburg/CN=radiora3-247625112e10-server
   i:/C=US/ST=Ohio/L=Cincinnati/O=Lutron Electronics Co., Inc./CN=radioRa3-processors-027-4725-46
-----BEGIN CERTIFICATE-----
MIIE+DCCAuCgAwIBAgICf4EwDQYJKoZIhvcNAQELBQAwgYIxCzAJBgNVBAYTAlVT
MQ0wCwYDVQQIEwRPaGlvMRMwEQYDVQQHEwpDaW5jaW5uYXRpMSUwIwYDVQQKExxM
dXRyb24gRWxlY3Ryb25pY3MgQ28uLCBJbmMuMSgwJgYDVQQDEx9yYWRpb1JhMy1w
cm9jZXNzb3JzLTAyNy00NzI1LTQ2MB4XDTIxMTIyODE3NDAxMloXDTQxMTIyODE3
NDAxMlowgYgxCzAJBgNVBAYTAlVTMRUwEwYDVQQIDAxQZW5uc3lsdmFuaWExJTAj
BgNVBAoMHEx1dHJvbiBFbGVjdHJvbmljcyBDby4sIEluYy4xFDASBgNVBAcMC0Nv
b3BlcnNidXJnMSUwIwYDVQQDDBxyYWRpb3JhMy0yNDc2MjUxMTJlMTAtc2VydmVy
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE5IGeB5PTdpHQMixJEQPX2UHPH71Z
Oyi6nRKtFz6dYOst29knfitTuCHq3r/RS2Wz/6dXtawTkYPVwTepTQVAuKOCATkw
ggE1MAkGA1UdEwQCMAAwHQYDVR0OBBYEFJJhe49OCQsytp+3ThHS99nz1yPkMIG6
BgNVHSMEgbIwga+AFLpN5eARwz9BeHCikhcb/9txVz2coYGApH4wfDELMAkGA1UE
BhMCVVMxFTATBgNVBAgMDFBlbm5zeWx2YW5pYTEUMBIGA1UEBwwLQ29vcGVyc2J1
cmcxJTAjBgNVBAoMHEx1dHJvbiBFbGVjdHJvbmljcyBDby4sIEluYy4xGTAXBgNV
BAMMEHJhZGlvUmEzLWRldmljZXOCFB43M/IDioTzc03ZsN39PEMvrCzNMA4GA1Ud
DwEB/wQEAwIFoDATBgNVHSUEDDAKBggrBgEFBQcDATAnBgNVHREEIDAeghxyYWRp
b3JhMy0yNDc2MjUxMTJlMTAtc2VydmVyMA0GCSqGSIb3DQEBCwUAA4ICAQAKoiFS
orwOo0RjRTyK5DHWOp8dS7la4H/mnjh77l5zpYOOjSrlQ82AANaQwQ3MbaGMupb5
g8pDMLcvf+3q6K9fii9h9+eCkrLCdQf16DInYqfOuRlqwCfNoxx+Q9SIMeGgO2pv
kHTf5j23nVmTBqSIVdNbMgFsa7nm30QmHyNTS/kh17dpi5MX2IyoW4rdv83Z7DXa
XznHHG0rgtipnTQH7KAKLX/+BtgfZg6r5GpmEPIX1sXmqpFzmmJtsVZ5cGLnA8uS
M4mU0SpdWZnv+5uFPqxHAYlRTkjc5QNoY7P3IQN4divYdjAt8y7AP42k6YQheM1i
fjfD+4I5X7ub7Uw8pBG5yxrW6CFF4ImZXxSgXhd0sHLqOWsz+zcyXeqF3WWEzXTd
lQA+S4BZQjJxNLaoyA+IMwk/lmTRUvtqPmOOYw7oRzeBH22OrMW7aTQyXlhanh3K
yHPniljS8ox5IzX+GIg2XHpaBDCE3HstlEc+MsX3LUGb07ZFXwTSO4LnhxTl1ql6
5blqsDEgeV5xbnckvHaKRREg7swszBFBXl9NL/nIEMAYE7zcnM195yT9th+yxXbj
IkxcwpMOagnwxD3EGlsw0G0qhsWapY2DrhXpjoW1kN4UXlwi3iasDDT3ATINtVas
Q7P8+3+B9AwVfPajfQQfZN7AgA1MMtTOoOavBQ==
-----END CERTIFICATE-----
 1 s:/C=US/ST=Ohio/L=Cincinnati/O=Lutron Electronics Co., Inc./CN=radioRa3-processors-027-4725-46
   i:/C=US/ST=Pennsylvania/L=Coopersburg/O=Lutron Electronics Co., Inc./CN=radioRa3-devices
-----BEGIN CERTIFICATE-----
MIIF8zCCA9ugAwIBAgIUHjcz8gOKhPNzTdmw3f08Qy+sLM0wDQYJKoZIhvcNAQEM
BQAwfDELMAkGA1UEBhMCVVMxFTATBgNVBAgMDFBlbm5zeWx2YW5pYTEUMBIGA1UE
BwwLQ29vcGVyc2J1cmcxJTAjBgNVBAoMHEx1dHJvbiBFbGVjdHJvbmljcyBDby4s
IEluYy4xGTAXBgNVBAMMEHJhZGlvUmEzLWRldmljZXMwHhcNMjEwMTA1MTcyMzA3
WhcNNDAxMjMxMTcyMzA3WjCBgjELMAkGA1UEBhMCVVMxDTALBgNVBAgTBE9oaW8x
EzARBgNVBAcTCkNpbmNpbm5hdGkxJTAjBgNVBAoTHEx1dHJvbiBFbGVjdHJvbmlj
cyBDby4sIEluYy4xKDAmBgNVBAMTH3JhZGlvUmEzLXByb2Nlc3NvcnMtMDI3LTQ3
MjUtNDYwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQDyTA9ITqOEV5X+
Y4JfzVdXPK0DscNyl/+WoX6hYsApKGMmKzYK7J7QwwiJ1plLPebAYYBityj5AkPV
6+Q3xAVkES1zpyIrCxdQPqUYEHj/TgdX98aShcyLG6h46WHnBglg/B6sGGHmn3lC
80VVcLh/2g4ArOkRvG/3n83gTGlypi2vU9src2a2iuKEriGhhGbfs1K0wzk2Ncht
uUdqYgeNSUrbDdznHnkxE2OMG0AZbUx4HuvD6ceuJQHcZTLG0Vz6QdG9w1IaPKma
zdyHvDifZEilqxGu31b7Tu9QKZRFGkx6W/4V1cn7zYhied6T9XCfiCXCFp8NklEy
hCRI1rI/zuMCUUq6lHs1DEPYxDoexwjg8r0fXvQKB/7IEyExpr6fuZbk9u5XTLna
2Qagab8xao+8i9ynGbW2W0PUGqV7ig1KrA9jqYVlwYd01nqhR/uxbZRbAdMH1CQ9
8PW6aD/MvcBqe2tZEsic8YWHfmk8g9V45ylbwZ5KyFD0P0DYCHYSHyRqQa//ovjx
s9CKacH9u967V80sNzq10FwU727epPy4QrQkPwbHk5Ax/2X3k1R8TaNpjPr7CTEt
+wQxVMBuI4BadCZcHbZbsqJvMr2FgARAMasgmaoVi66QLE6H/btPlZNLMWaLCWjC
KlrFb1gz15q4jeLJFxi3PSUYgmUfgQIDAQABo2YwZDAdBgNVHQ4EFgQUuk3l4BHD
P0F4cKKSFxv/23FXPZwwHwYDVR0jBBgwFoAUNQ8++bdtiOuIkXJQh7r13+CWc8Uw
EgYDVR0TAQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAYYwDQYJKoZIhvcNAQEM
BQADggIBABwtHHRXpJR2WXu37IM0Ik50cc1oICYtSTJpb/Lfuj4Wo7hS21519Sca
FwoOcnU1sKlOsueMJRKTz05wUiS+vwKGMtYXHU5b1aCBsRnP45+Xel9She2sN86U
wWkrnRPJ4gzLiqYovoDAA5WoG4n5zHQb4+I7s2lJs9e1jlUZ7eLJnVW8+lWfSwVH
9K0QFvxveVqzyhX+zQ0nK3tvVjN33iba6Ao8qX2CdPzOF6/f3uHsBzSD+NbFofZO
5G2lKSKwBQogsWBx1KEbJBXq53mTux5MqengppemZ+NMDSjT7/BeM1MidBSuks6J
L74EVt33RYkJKEk/uGbQPR3yO430BZgFaQRFOsGwYzr1ls009+cz4jCG4KGXkpEZ
pCjIAq0CKgEqFYVBUZEjzSw+xQcGTsqLSCkFm8GzCx6ni4jUeFScOBZGz6KIcPbG
h8Prxa415jLVb3fcXhIK89h+EkRs0t4Gl2ZmVUTWpJbCVlRhmWm+jh0xtMQ/xlak
2gXHHDRYSBwFRNaakEr2q9wGUE0y7nLO2w5Er/5jPCgfPWQGmmzFfENJoW45bUfo
o6VGnF/JppAGUgWKwJvh/J8nlPbfL/KHcN6AmucxenMu7eqH5um5334xo1NzsQ+k
jHXWqunPoo3i6RKjsuUqh7QOgrTgI2T+Sman0mbERsrYMV7xMgk7
-----END CERTIFICATE-----
 2 s:/C=US/ST=Pennsylvania/L=Coopersburg/O=Lutron Electronics Co., Inc./CN=radioRa3-devices
   i:/C=US/ST=Pennsylvania/L=Coopersburg/O=Lutron Electronics Co., Inc./CN=radioRa3-products
-----BEGIN CERTIFICATE-----
MIIF7TCCA9WgAwIBAgIUF6a856xOgdHLwool4dRPBmqKNwUwDQYJKoZIhvcNAQEM
BQAwfTELMAkGA1UEBhMCVVMxFTATBgNVBAgMDFBlbm5zeWx2YW5pYTEUMBIGA1UE
BwwLQ29vcGVyc2J1cmcxJTAjBgNVBAoMHEx1dHJvbiBFbGVjdHJvbmljcyBDby4s
IEluYy4xGjAYBgNVBAMMEXJhZGlvUmEzLXByb2R1Y3RzMB4XDTIwMDgxNzIxMzYz
MFoXDTQwMDgxMjIxMzYzMFowfDELMAkGA1UEBhMCVVMxFTATBgNVBAgMDFBlbm5z
eWx2YW5pYTEUMBIGA1UEBwwLQ29vcGVyc2J1cmcxJTAjBgNVBAoMHEx1dHJvbiBF
bGVjdHJvbmljcyBDby4sIEluYy4xGTAXBgNVBAMMEHJhZGlvUmEzLWRldmljZXMw
ggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQDBYzHw0ljz63WYtV5ZnHTU
J75y7e6dp2yvBIqksZPBWGkwO0WYQgUsDhEYA1bS3Es+KM3zgENGTIUlcG0a14+c
bs2uxmzPWICK2yXg0OxZmzqqVFSCbBuW1d/VCw+dTxGml2f3/grWFakqc/XB6BpK
SokzWqZcxbUSrn8Znsrvp8x8R6JaGlT7tRZ6EaGksTYRLOHZ43/mVnFYrlsxkeaD
lQFjJenyg5Z1pFAXFTLlCbdl59vphj8iQjJuEgFnyK9EFOgiGSd9dT3xoNKerfmy
DosxTEvtkXfLT1XegVkEm/z9/W4zBKfOa1fWCAB07/vfsxsSg2efvXvPq9IYHI0Y
PQLRlV+usEqhvksZ7BVPzIaN8VRmZ94x+8vvtv//Hc02pKYdESXW63qrAckmUst9
fZGVgcowrLSJC5jeCGmXsROdTi8Qo9UDbNsCS/RlQaFkJlgRLpW+KuScNqm8Tq7B
rUZS3Yp5Y9MhGFN0BSP12JpMpSDO1liRR6kDXx+dMi7IYclUt6a3SKhV79bA1jNN
e0B4NJVehfdANQg/9vJj5Jj4mFbh9GZ7r/9OXgfmuvS9NjeGf5iD3/Tt1KIh9Svn
t8sD7CPngWuVm6jht/+H9U/IxPY1jLZJP4+XAyfsJCIQ6pUHgP6fm5njtJJO5p0o
OVWHvhg3QyN98VNyjH66AQIDAQABo2YwZDAdBgNVHQ4EFgQUNQ8++bdtiOuIkXJQ
h7r13+CWc8UwHwYDVR0jBBgwFoAULo6eKrhJfxt9XmmKsqIVHTXUz8IwEgYDVR0T
AQH/BAgwBgEB/wIBAjAOBgNVHQ8BAf8EBAMCAYYwDQYJKoZIhvcNAQEMBQADggIB
AJWQdI0sb2FR+j9OSNq8wGTZ9p4kWt1QVFOzCG14EOdBZU2BTlUmM9ICTRbt1Czs
AfHUBwA5Qj0vamOfsrlnfbDtwAjbO+t10MOkaqa8w9/jP7/VaqzBGruI5XEbvHXd
kr4YMx/sY53QM7SewXQrC7pIk7PAIOhNE2HmqcgXH2u9azX9Rp8Zgsn2WVULeIz0
EmTJygpwCTUBTmJRgMHQwDH+ZP8yhv2CgTTdtFVI29vog6wQi1PzgjFvqWBJKyTa
fzIdHhcxSTbcLl+d1j5yYGXQvyvvJoGMaKSmfabOhnHR7mqTXXZmZjuC2FHjSQa8
4FR0VBEFbyZoXpbeIybabiFM7SmD1/W7UEfmgHLyj7F85rTgjL3Sp3w7hjlsHCvC
5xfXFVfWax0T8V4gN1dpxI2m3NA47nND5j6lHJmj3P6Wki1GsM1gJ/FlM0ipge11
7pRifRm5WNeXhZX7ewX87fmGJdbAFWLjEzr+IenOmQU7UdSqRvN58RpDjYsuuhSN
02F6ZFsr25pG0vgM5y/J/vV74TLkGc1xbqFSFvBEhZmzaUsjWiz/MbWdY+/wxmQI
auvS9iFUBO3My+N8sUSEuYYOi0gxGCCr0OXaYES+l1uVdVs5m1SD3km6O7hBEGuN
GTtHwnoDD+BKARmHrw9lQbFL4nkMb4c/tNsz4oUehXxp
-----END CERTIFICATE-----
 3 s:/C=US/ST=Pennsylvania/L=Coopersburg/O=Lutron Electronics Co., Inc./CN=radioRa3-products
   i:/C=US/ST=PA/L=CB/O=Lutron Electronics Inc/OU=Lutron Electronics Inc/CN=lutron-root
-----BEGIN CERTIFICATE-----
MIIG8DCCA9igAwIBAgIUFRRjHKJHQM8Iibgy5fv3mGqiBPAwDQYJKoZIhvcNAQEN
BQAwfzELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAlBBMQswCQYDVQQHDAJDQjEfMB0G
A1UECgwWTHV0cm9uIEVsZWN0cm9uaWNzIEluYzEfMB0GA1UECwwWTHV0cm9uIEVs
ZWN0cm9uaWNzIEluYzEUMBIGA1UEAwwLbHV0cm9uLXJvb3QwHhcNMjAwODE3MjEz
NjMwWhcNNDAwODEyMjEzNjMwWjB9MQswCQYDVQQGEwJVUzEVMBMGA1UECAwMUGVu
bnN5bHZhbmlhMRQwEgYDVQQHDAtDb29wZXJzYnVyZzElMCMGA1UECgwcTHV0cm9u
IEVsZWN0cm9uaWNzIENvLiwgSW5jLjEaMBgGA1UEAwwRcmFkaW9SYTMtcHJvZHVj
dHMwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQDyXA8T4esN7gzz05fN
+KQosUzh+A4xMaEmpINY2fRl+iSybvEizjLnCSnLt3PQ6cD+NRmYv4InLPX71Ew6
ngznMVKsE7yzEFLAZynA+1Miy7Ivz5SsltB/Ps9XPLpKf9mLMJ4aP00g8NAbwfu/
yio7I8ZGvO3d9BxxwQ3ZRduZyzglVW1BrzVAA6WMgzlPuyHUWxBl/bnwYg0HBdgh
WZyI9eGxC50N2oFfW2089WjRfZasgl6x9+Jji7O32fKXprPQL2HobxpGNUoqs4tS
PihwZZp3xd8XLVsonNFtAWZ0ukOiaCtGK38Z44/OL2oZLaPJAj61HwivQcFUz3Qn
fe7dqQuoqELlSBp8rk08UAZpdLX1n6TpEwhlND8f1/fQWOnMl0yvZ9mkUIqlF9ua
QNZuYKpjLKBlJnCpg5zandRTNk4b7Z8CIXe02nPvBxePy0BHrwK8sQdxPkVgzeox
vTfrw/2I425gABQeNb0fICtLRuxqoJvgHGHieeQTZP1KEUHPBFGWrr8za/dnXzQC
/4pUK2S1pdY6tL7vLnZ9izdCjcLRuhYLSaaXblyLoQSTNPuHGRupcfW4j0ESFrGY
sWkJ6xtwxWcnUsgZLzqlUEvxKkFN1MAAyEGVrscNf15DEZc9+xzK3yETak/ab7oE
gjDfHLJx5wBOcyr0L26QSnMEiwIDAQABo2YwZDAdBgNVHQ4EFgQULo6eKrhJfxt9
XmmKsqIVHTXUz8IwHwYDVR0jBBgwFoAUStvZKYaSH7iRh7a3BKJ7n7FqpGwwEgYD
VR0TAQH/BAgwBgEB/wIBAzAOBgNVHQ8BAf8EBAMCAYYwDQYJKoZIhvcNAQENBQAD
ggMBAJmURob8hLsYRH2BsSfrRlRokDmqD7+zXCR3T7SfRXETLdxyciqKSg93+rry
VJT2WjnKvsPc4V1X34k2mzC+4D24O/9yKRvKk/Jdq2usuelePtURSoD8X+lkJJ5F
2sOfZCcZ1D6XMlyfBS3oZofm4MfL5Pftv5xLWRg/gdNDbWOwVcGmxMjAJEwpDrCs
fO1SII03r3ImGa1fayfvzCoP1DX2QgytMqfmtsPljyE4YqIRp7/Rrv39JaqY3lOb
PvmAx4DtAkISeOjYYu1RFuOE7Cyw6t2zBVISj1KVQZ6btEeTkROpYGDNHOl5Ke8o
aUkmH5ShHujIrz1f1ylg2jUmWdin1MUtmAVNuPgx7lQE5sLjZS7/UO/A4De+ECsA
t4McOF+6RS5N86u8Y80MFzybpI3WnpqLMmdbYw7xwuLcZB+OCUUNwn6sxAn4eHiq
WhLB5wQ4nHr4LfRqyMHQ7s77TuDkTSO51tUbBbCu0E0rCxeV1I3EBErk7quM297+
2qnvW5k83DfI0qfRI/dAz6jhdFUx3733merauVY8NpTY6h6OoVBEW7HYJC/+Ju7Q
wu008tatR3wo+Cj25JsWCiJSfKh27N0IgrrnHrpPcXpLneiv5U0pjgLGcyh2AWC9
ERJF3v3f1JuIJb/H5jVlLAh+oI9JjjNYj891JSDmomP8Gp6h51IE68Q5xhg3J6Es
jSi+xqXMvfIbHcuTaGFw03B1aXtPgr1HvLx2RX3fGkdUM8PCA2e4uK/1aKjfXuIU
QMy9wEmz//XecE71C4R8abvrDcmjK4kRH/y9LZGOSh41iWM1Y52fqEAuRz4IK4//
9Il0dnxBv/atyQlR41E64bLF2zl3FfeF8DM1qYWTZxaqMf4qf0AjssNa/NIV46cN
Nssil5YIMauvJfvFhJEdK5AbVcMhvjX9iCXScZS/sBoCjZ8OF65sACW/a3amwLzD
i2HfWHC6DVhNsA4hAD1YtHFb6mPfSARmQFrwtxQLQRoSgtd3Wyo2Q5L80joGSkJ9
qU6/xw==
-----END CERTIFICATE-----
 4 s:/C=US/ST=PA/L=CB/O=Lutron Electronics Inc/OU=Lutron Electronics Inc/CN=lutron-root
   i:/C=US/ST=PA/L=CB/O=Lutron Electronics Inc/OU=Lutron Electronics Inc/CN=lutron-root
-----BEGIN CERTIFICATE-----
MIIH5DCCBMygAwIBAgIJAKk++JqaJetSMA0GCSqGSIb3DQEBCwUAMH8xCzAJBgNV
BAYTAlVTMQswCQYDVQQIDAJQQTELMAkGA1UEBwwCQ0IxHzAdBgNVBAoMFkx1dHJv
biBFbGVjdHJvbmljcyBJbmMxHzAdBgNVBAsMFkx1dHJvbiBFbGVjdHJvbmljcyBJ
bmMxFDASBgNVBAMMC2x1dHJvbi1yb290MB4XDTE2MDkyODE5NTk0MVoXDTM2MDky
MzE5NTk0MVowfzELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAlBBMQswCQYDVQQHDAJD
QjEfMB0GA1UECgwWTHV0cm9uIEVsZWN0cm9uaWNzIEluYzEfMB0GA1UECwwWTHV0
cm9uIEVsZWN0cm9uaWNzIEluYzEUMBIGA1UEAwwLbHV0cm9uLXJvb3QwggMiMA0G
CSqGSIb3DQEBAQUAA4IDDwAwggMKAoIDAQDBZbMODMzm+qpsOF5hhQ272GUlOaKz
n5b5YxokSAoxY4TqQApb9/uRHIBuuGLntq0QhR0Y3b0lXBeJWzWC6zscZJrheUKW
+2aHVvU4ugPAAXK/WVI68adBSY1UP0BcO1paYrXONcuXQgdy2/GV1mo1b+bmjNFT
zeDopkUoBxivBDZZ7B5vFfbJSgSF47Xsz8cspCEUIaV1rZbaDYBzsimdvrKusJfZ
Pci+Cx71sZuKunGTCgwHduYFsBfYRgTG1ihNEASi2++Er67AcabUGaqVQr/kIrUD
sS9jB6uaqPgMajjwXiZPDm82tTHobbKSav7aq+kSBNIFyvhK5y+vAWoGeZr5WK7n
9EekO3x7LXc6XSCASuhzK6zquAGUBSQNEO3c7sZ1rIdNs1lBSkCSxs+Bl8eEHO8k
O20TqKzKF9bQtccNkFWtRKIhVLFxQt234P+XJtWvWKVOlkLCAo0QgDivFJQVnNKM
Hr2/CIsOLC+ZSWAYl0lZEJaszt7wjR9cc7DRizq9aoKcGlPRvxzobFoQ4H0Z8vIR
DQRUQWFaTTOGiEk7JKxqjXX8xuGZpoXWw8VX0gz3Y0Bz8sU58ZZbugmVjvnKKYzd
ueZ/9+FsaYX6CKdJDANEJf+fqfkGXwQGt8Ns7SeG0JyCdJ4K2ECoOURYS4P1vSY1
40L9OldDjsW2qhpSBPHppfJ4rPRUu5J9Ux3AX4Mz+ibl6MS3wRpP1Rg+9TLITK5o
3AYrJO6oMsYrQkQvc+k20ocD7Iq0522iyw62/DpKMsPZXHNTT/rqzIihkaZaR8aa
ZOgAKi5o398mcfsuv42f8DriYc0Gr+3btiDU7rINqM935YNIABBDtVT9Ybc5uPHa
wXLmAIx2yLjqYaRDhr01Sql6WGy8Y0HcI5lM1pw4Vpx+VKWG/QdORGtZgKySGZ0+
9bY9cRN9IBFz4J60xoqx0MsM5o6FqVDypDCB32KaobVZSAnHifwEGtJJimNIzHpY
jGCTzBHSpuZcvV2dVAuPTHzck37ifpNTUFUCAwEAAaNjMGEwHQYDVR0OBBYEFErb
2SmGkh+4kYe2twSie5+xaqRsMB8GA1UdIwQYMBaAFErb2SmGkh+4kYe2twSie5+x
aqRsMA8GA1UdEwEB/wQFMAMBAf8wDgYDVR0PAQH/BAQDAgGGMA0GCSqGSIb3DQEB
CwUAA4IDAQAP9t0STrn0xENQEDNwRMQUBYTA/r7BmzXX6b0Ip6HW7QNmTkFc5pUb
KT86v6g8cJ7ls96JwdTu1kRQt4Qpbyvzp2p5YlRFnm3NTVVdcffcZNo95x9Z/1Cv
5xZgw0OKODwPBJLCyq5ET4W6WrIZucRVBs035YXIN+z3EzCxBj6O1lcjOTTHSFFR
jI3t3AGdkFo7tCBu5TFlNEFfmaqjse140vfGWJpRKyOT4ahzXLcVxzfg/SKRID3Q
2Rop4KqLNCddzz+UM+IiwyFkOfjrXWStW46cLUzM1k5GRrl0aBg5oqMCBY4/Eeh3
W0ZATsxxfg2Ly4FIO5d7/xiZqARFuurYe/2PSzVSPIKQrPVjDEekD+qQ0bRRQGvL
KBiMhYZHwnz/OEQl21PNp5rksuFjKG4/PimQ8V96jpbzzsOZuic3aScszgNUPbdI
0LYjCQ8xCOFPpFC4x1+rGubRjKEuGCvvYErVkX9rQlFRGPOp+k8bYTlIUKZeNsuL
KiZ4VH5+ZUAIf94DHayoo/SvBsQ5Qizb17KVRKil+vidUkMtndrNtjr3GWmH+nkn
PhRXBlekUy3dgRvTE8RFDOG9TYAN1Bs/uMgNc8Sg5Yz0SG96SLXVer2zsjmQ7tf9
6s+UVvrr+wlL7jSJCfJo6gaUQh1sD3umPXDS+Fq+J7tiRwOvP3cejo8dLyhesDun
FGIHlKmUCIwS/3Kzvd9OtAJMsmV9Q2B1dXudJloj6ADaAmVvhI/eMUncL9sXMJZH
3CCorh2OSZt0vtdA59osgDSMrsQZSMrtovrKgeFmP1Z0ENvo90Zenm7Bjn6Hw3Y/
GebIgSgoc149ElxjN4nagIqSJJHRrYq85sjTUSESvQUL1oi4R/VU+qMIRSHju/ZM
bkqONDohUc7/pg5rnLTZnnaQ09KvdF0yySx3hYph7L7MZWV/tF7O7yj1egRKh7lT
rgZOI7EiN4DPfTTpXoWVmIpiB/ouKp6uZ/Zrq00WthT8lUaBsFYaC3FDkkcxwdkk
lJ+cvdbUdsU=
-----END CERTIFICATE-----
---
Server certificate
subject=/C=US/ST=Pennsylvania/O=Lutron Electronics Co., Inc./L=Coopersburg/CN=radiora3-247625112e10-server
issuer=/C=US/ST=Ohio/L=Cincinnati/O=Lutron Electronics Co., Inc./CN=radioRa3-processors-027-4725-46
---
Acceptable client certificate CA names
/C=US/ST=Pennsylvania/L=Coopersburg/O=Lutron Electronics Co., Inc./CN=RadioRa3Processor247625112E10
/C=US/ST=Pennsylvania/L=Coopersburg/O=Lutron Electronics Co., Inc./CN=Installer-RadioRa3Processor247625112E10
/L=Coopersburg/O=Lutron Electronics Co., Inc./CN=Lutron Project SubSystem Certificate Authority/ST=PA/C=US
Server Temp Key: ECDH, X25519, 253 bits
---
SSL handshake has read 9017 bytes and written 301 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-ECDSA-AES256-GCM-SHA384
Server public key is 256 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-ECDSA-AES256-GCM-SHA384
    Session-ID: 8FBD45527656870F331D560C75F201CC6A0EF0118598DC1B1DF0EC64D25DC9B1
    Session-ID-ctx:
    Master-Key: 05360F18DAEB8AB3C2C005BA1F86817E33C1CC364F14130974A4C56A456BC949668C76883163A3C8469102F023F06CFD
    TLS session ticket:
    0000 - 27 cb 07 fb a2 cf 75 9e-9b 5c 3c 7b 9a 32 7f f8   '.....u..\<{.2..
    0010 - 5f 0b 3e a4 c4 8e 0c e3-82 20 a0 bf 96 4f 3b 0d   _.>...... ...O;.
    0020 - eb 3e 5c bf 5a 55 d6 a9-4b c9 3c 51 30 ac 62 e4   .>\.ZU..K.<Q0.b.
    0030 - 91 05 a3 7b a8 f6 ff 3c-3a c1 38 34 f5 82 7e e4   ...{...<:.84..~.
    0040 - fb 60 35 0b 50 be 46 42-61 31 ad 01 f1 53 d3 b6   .`5.P.FBa1...S..
    0050 - 7f 7c 88 ac 8e 89 59 b6-78 69 db 81 f9 4d c7 f6   .|....Y.xi...M..
    0060 - a4 30 e7 db 8e fc 16 64-08 db 97 a7 ca 1f cb 7e   .0.....d.......~
    0070 - f0 8b c7 40 aa ba ca 8c-                          ...@....

    Start Time: 1644972097
    Timeout   : 7200 (sec)
    Verify return code: 19 (self signed certificate in certificate chain)
---
```

Reply:
That's interesting. For the service they use a different certificate chain where the bridge isn't a CA. Try using just the `C=US/ST=PA/L=CB/O=Lutron Electronics Inc/OU=Lutron Electronics Inc/CN=lutron-root` certificate. Everybody's RA3 certificate should have the same one so it could just be included in the code.

pylutron_caseta could have a copy of lutron-root and then when connecting to a bridge it could verify that `radiora3-247625112e10-server` SAN record matches the expected value. That's probably what official software does. I suspect this change was made to simplify integrations. If another device on the network has the lutron-root certificate and knows the same of the bridge it's connecting to then the security works about the same as common TLS.

But I don't know if it's worth messing with. Since the leaf certificate is good for 20 years it's probably fine to just save the current leaf certificate during the pairing process (the rest of the chain shouldn't be necessary). Verifying that the certificate chain connects to lutron-root could make man in the middle attacks during the pairing process much more difficult. I don't think that's a big concern. Saving the leaf certificate should work for both Caseta and RadioRA3. Saving the full chain and not validating the subject name creates a new potential man in the middle where somebody compromises a different bridge to get a certificate signed by lutron-root and then impersonates your bridge.

Also, I think for RadioRA3, pylutron_caseta should _not_ trust the bridge client CA certificate (caseta-bridge.crt) because it looks like the certificate issued to you when you pair is valid for server authentication. That could create the possibility for a different man in the middle attack if pylutron_caseta is not validating the expected subject of the target device because then it would trust any other device which has paired with the bridge as if it were the bridge. (I'm not sure what anybody would have to gain from doing this since they already have access to authenticate to the bridge)

TLDR:
The client root CA (currently caseta-bridge.crt) should probably be concatenated onto the end of the client certificate (caseta.crt) instead so it's only used for authenticating pylutron_caseta to the bridge.
The leaf certificate of the bridge during the pairing process should be saved as the server root CA (new caseta-bridge.crt).
This should work for both Caseta and RadioRA3 and provide good security for both, at least after pairing has completed.

Reply:
For RA3, putting the mentioned lutron-root cert in the caseta-bridge.crt alone with no other certs, and the caseta.crt also alone with no other certs works just fine.  The data that used to be in caseta-bridge.crt is effectively gone with no ill effects.

But with Caseta, this isn't the case.  If I concatenate the old caseta-bridge.crt into caseta.crt, and use the lutron-root cert in the caseta-bridge.crt file, the client still doesn't trust the server (whose cert matches the server cert, so it's necessary for trust). 

It seems maybe we need to handle it two different ways?

For Caseta, leave it as is, since the client needs to trust the server using the caseta-bridge.crt that matches the server running on 8081

For RA3, all left as is, but replace the caseta-bridge.crt with the lutron-root entry (static & hardcoded?).

Does this seem right to you?

Reply:
I think we should be able to use just the leaf certificate, in your case radiora3-247625112e10-server. With Caseta the leaf certificate is the same as the CA certificate so the same code would work for both.

Reply:
With the normal caseta.crt and the caseta-bridge.crt containing only the radiora3-247625112e10-server cert, it fails.

```ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1129)```

It seems without the root self-signed cert in the CAFile, it's not happy...

Reply:
The client needs the root, and the server provides the full chain so the client can validate trust... so I was wrong to include the full chain when I only needed the root.  

But it seems we'll need the root cert for caseta (which has no chain), and the root cert for RA3. If I always choose the last cert in the chain of trust (root), it should work correctly, no?

Reply:
In summary:
In Caseta, none of the chain certs work.  It only works when using the normal/typical caseta-bridge.crt
In RA3, the typical returned cert fails, and it only works when providing the root cert from the chain.

Can you confirm this with your local Caseta?

Reply:
Not working with just the leaf sounds like one of those weird Python openssl things to me. Maybe there is a way to make it work, but I remember there being issues before where different versions of Python on different computers had different behaviors in some more obscure cases so maybe it's not worth trying to get it to work that way.

What do you get if you make this LAP request on port 8083?

```json
{
    "Header": {
        "RequestType": "Read",
        "Url": "/certificate/root",
        "ClientTag": "get-cert",
    }
}
```

Does it give you the same certificate as pairing or does it give you something like lutron-root? For Caseta it is the same so pylutron_caseta doesn't bother getting it. Reviewing the LAP stuff, it really wouldn't make sense for there not to be a way to learn during the LAP phase what certificate to expect when connecting to LEAP.

Reply:
Same certificate as pairing, unfortunately.